File: javascript/src/problems1_100/minimum-window-substring.js
class Problem {
  minWindow(s, t) {
    throw new Error(`Not implemented (${s}, ${t}).`);
  }
}

class Default extends Problem {
  minWindow(s, t) {
    if (t.length > s.length) {
      return '';
    }

    const sMap = {};
    const tMap = {};
    for (const c of t) {
      tMap[c] = (tMap[c] || 0) + 1;
    }

    let left = 0;
    let right = 0;
    let result = '';
    while (right < s.length) {
      let c = s[right];
      sMap[c] = (sMap[c] || 0) + 1;
      right++;
      while (left <= right && this.contains(sMap, tMap)) {
        result = this.getMinLength(result, s.slice(left, right));
        sMap[s[left]]--;
        left++;
      }
    }
    return result;
  }

  getMinLength(s1, s2) {
    return s1 === '' || s1.length > s2.length
        ? s2
        : s1;
  }

  contains(sMap, tMap) {
    for (let key in tMap) {
      if ((sMap[key] || 0) < tMap[key]) {
        return false;
      }
    }
    return true;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems1_100/minimum_window_substring.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def min_window(self, s: str, t: str) -> str:
        pass


class Default(Problem):
    @override
    def min_window(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ''

        s_map: list[int] = [0] * 128
        t_map: list[int] = [0] * 128
        for c in t:
            t_map[ord(c)] += 1

        left: int = 0
        right: int = 0
        result: str = ''
        while right < len(s):
            s_map[ord(s[right])] += 1
            right += 1
            while left <= right and self.contains(s_map, t_map):
                current_substring = s[left:right]
                result = self.get_min_length(result, current_substring)
                s_map[ord(s[left])] -= 1
                left += 1
        return result

    def get_min_length(self, s1: str, s2: str) -> str:
        if len(s1) == 0 or len(s1) > len(s2):
            return s2
        return s1

    def contains(self, s_map: list[int], t_map: list[int]) -> bool:
        for i, m in enumerate(t_map):
            if m > s_map[i]:
                return False
        return True


SOLUTIONS: list[Problem] = [Default()]
